### DISCUSSION TITLE
web page for my science fair project

### DISCUSSION THREAD

**Discussion by ethansopro** [pybricks/support#1983] [Q&A]
I am working on this projects for around a month and i need help makeing a webpage that has 3 buttons (i am making a drink dispenser) I already got the interface already( i am using a technic hub and using vs code and pybricks to try to make it work but i cant(
![Screenshot 2024-12-20 215736](https://github.com/user-attachments/assets/bf79c9b6-cd67-48b8-8180-4ce4c05e2a01) 
the thing is i cant make it work i worked really hard and if any one can give me intructions and how to set it up and the code , i will grately appriciate it thank you

**Comment by dlech** [comment]
Can you share the code that you wrote already?

**Reply by ethansopro** [reply]
sure 

```python
import threading
from flask import Flask, render_template, request, jsonify
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Port
from pybricks.tools import wait
from googletrans import Translator
import os

# Initialize Flask app
app = Flask(__name__)

# Initialize Google Translator
translator = Translator()

# Initialize Technic Hub and Motors
print("Connecting to Pybricks Hub...")
try:
    hub = TechnicHub()  # Automatically connects via Bluetooth
    print("Technic Hub connected successfully!")

    # Initialize motors
    motor_sprite = Motor(Port.A)
    motor_coke = Motor(Port.B)
    motor_fanta = Motor(Port.C)
except Exception as e:
    print(f"Failed to connect to Technic Hub: {e}")
    os._exit(1)

# Serve the HTML page
@app.route('/')
def index():
    return render_template('index.html')

# Endpoint to handle drink orders
@app.route('/order', methods=['POST'])
def order():
    try:
        data = request.json
        drink = data.get('drink')
        lang = data.get('lang')

        if not drink or not lang:
            return jsonify({"error": "Missing 'drink' or 'lang' parameter"}), 400

        # Translate message
        translation = translator.translate(f"You ordered {drink}", dest=lang)
        translated_message = translation.text

        # Dispense the selected drink
        threading.Thread(target=dispense_drink, args=(drink,)).start()

        return jsonify({"translated": translated_message})
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Function to control motors
def dispense_drink(drink):
    try:
        if drink == "Sprite":
            print("Dispensing Sprite...")
            motor_sprite.run_angle(500, 180)  # Run motor at 500°/s for 180°
        elif drink == "Coke":
            print("Dispensing Coke...")
            motor_coke.run_angle(500, 180)
        elif drink == "Fanta":
            print("Dispensing Fanta...")
            motor_fanta.run_angle(500, 180)
        else:
            print("Unknown drink selected!")
    except Exception as e:
        print(f"Error dispensing drink: {e}")

if __name__ == "__main__":
    print("Starting Flask server...")
    app.run(debug=True, host="0.0.0.0", port=5000)
```

**Reply by dlech** [reply]
> ```python
> 
> from pybricks.hubs import TechnicHub
> from pybricks.pupdevices import Motor
> from pybricks.parameters import Port
> from pybricks.tools import wait
> ```

Everything in the `pybricks` package can only run on the hub, not on your computer. So you will need to make two programs. One that runs on the hub and the one you already have that runs on the computer. Then you will need to make those two programs communicate with each other.

There is a tutorial on this at https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-communication/

There is also an newer way to send info, but it doesn't have a lot of documentation yet. https://docs.pybricks.com/projects/pybricksdev/en/latest/api/ble/pybricks.html#pybricksdev.ble.pybricks.Command.WRITE_USER_PROGRAM_META

**Reply by ethansopro** [reply]
i already tried it on my v2 code
![image](https://github.com/user-attachments/assets/174d9767-2849-4587-b70a-fa68d78c90c4) and I will send you the zip file for it 
[DrinkDispenser (2).zip](https://github.com/user-attachments/files/18212543/DrinkDispenser.2.zip)
( i tried finding ways to do it by making it simple then adding features on the v1 code )

**Reply by dlech** [reply]
Looks like you are mostly on the right track. But I can see the problem. The `pybricksdev` process to run the remote program is getting called for each command. But this will cause the program on the hub to keep starting and stopping.

Instead, we need to make sure that `pybricksdev` is only run once when you web server starts (or from a "connect" button in the UI). Then communicate with that process from the web server. Unfortunately, https://github.com/pybricks/pybricksdev/issues/17 has never been implemented, so we can't use stdin of the `pybricksdev` process the way you are trying to.

So instead, you can use `pybricksdev` as a Python package instead of running it as a separate process. Since it uses asyncio, you will need to add another thread to your program that calls `asyncio.run()` to start the async event loop.

**Reply by ethansopro** [reply]
can you help me do it sir?

**Reply by ethansopro** [reply]
czn you giv me an example code witha red button if click motor a moves 720 degress and etc'
